pint/quantity.py

```diff
@@ -79,6 +78,262 @@ def wrapped(self, *args, **kwargs):
         return result
     return wrapped
 
+HANDLED_FUNCTIONS = {}
+
+def implements(numpy_function):
+    """Register an __array_function__ implementation for Quantity objects."""
+    def decorator(func):
+        HANDLED_FUNCTIONS[numpy_function] = func
+        return func
+    return decorator
+
+def _is_quantity_sequence(arg):
+    if iterable(arg) and sized(arg) and not isinstance(arg, string_types):
+        if isinstance(arg[0], Quantity):
+            if not all([isinstance(item, Quantity) for item in arg]):
+                raise TypeError("{} contains items that aren't Quantity type".format(arg))
+            return True
+    return False
+    
+def _get_first_input_units(args, kwargs={}):
+    args_combo = list(args)+list(kwargs.values())
+    out_units=None
+    for arg in args_combo:
+        if isinstance(arg, Quantity):
+            out_units = arg.units
+        elif _is_quantity_sequence(arg):
+            out_units = arg[0].units
+        if out_units is not None:
+            break
+    return out_units
+    
+def convert_to_consistent_units(pre_calc_units=None, *args, **kwargs):
```

Comment:
it looks like `pre_calc_units` should be a keyword (-only?) argument, since it's easy to forget to pass `None` if we want to keep the default